File: _vault/NFBak/Neuroflow_vsonline/Neuroflow/OldStuff/Neuroflow OpenCL/Neuroflow.Networks/Neural/CPU/LayerForwardCompute.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics.Contracts;
using Neuroflow.Core;

namespace Neuroflow.Networks.Neural.CPU
{
    public abstract class LayerForwardCompute : LayerForwardComputeBase
    {
        protected LayerForwardCompute(ConnectedLayer connectedLayer)
            : base(connectedLayer)
        {
            Contract.Requires(connectedLayer != null);
        }

        protected internal bool RunParallel { get; internal set; }



        protected internal virtual LayerBackwardCompute CreateBackwardCompute(ConnectedLayer connectedLayer)
        {
            Contract.Requires(connectedLayer != null);

            throw new InvalidOperationException("Backward computation is not supported.");
        }

        unsafe protected internal abstract void Compute(float* valueBuffer, bool collectTrainingData, int? innerIterationIndex);

        unsafe internal void Reset(float[] valueBuffer, NeuralNetworkResetTarget target)
        {
            if ((target & NeuralNetworkResetTarget.Outputs) != 0)
            {
                ValueBuffer.Zero(valueBuffer, OutputBuffer);
            }
            else if ((target & NeuralNetworkResetTarget.Ps) != 0)
            {
                if (Method == ForwardComputationMethod.RTLR)
                {
                    foreach (var range in PBiasBuffers)
                    {
                        ValueBuffer.Zero(valueBuffer, range);
                    }

                    foreach (var lvl1 in PWeightBuffers)
                    {
                        foreach (var lvl2 in lvl1)
                        {
                            foreach (var range in lvl2)
                            {
                                ValueBuffer.Zero(valueBuffer, range);
                            }
                        }
                    }
                }
            }
        }
    }
}

File: _vault/NFBak/Neuroflow_vsonline/Neuroflow/Neuroflow-SimplifiedCpp/Neuroflow.Native.Impl/LayerForwardCompute.cpp
#include "stdafx.h"
#include "LayerForwardCompute.h"
#include "CPPAmpValueBuffer.h"

using namespace concurrency;

LayerForwardCompute::LayerForwardCompute(void) : 
    connectedLayerIndex(0),
    isOutput(false),
    biasValueIndex(-1),
    method(ForwardComputationMethod::FeedForward),
    biasGradientValueIndex(-1),
    biasGradientSumValueIndex(-1)
{
}

void LayerForwardCompute::Reset(CPPAmpValueBuffer& buffer, NeuralNetworkResetTarget target)
{
    if (((int)target & (int)NeuralNetworkResetTarget::Outputs) != 0)
    {
        buffer.Zero(OutputBuffer);
    }
    else if (((int)target & (int)NeuralNetworkResetTarget::Ps) != 0)
    {
        if (Method == ForwardComputationMethod::RTLR)
        {
            for each (auto range in PBiasBuffers)
            {
                buffer.Zero(range);
            }

            for each (auto lvl1 in PWeightBuffers)
            {
                for each (auto lvl2 in lvl1)
                {
                    for each (auto range in lvl2)
                    {
                        buffer.Zero(range);
                    }
                }
            }
        }
    }
}
